catroid/src/main/java/org/catrobat/catroid/stage/PenActor.java

```diff
@@ -103,18 +111,20 @@ private void drawLinesForSprite(Sprite sprite) {
 		renderer.setColor(pen.penColor);
 		renderer.begin(ShapeRenderer.ShapeType.Filled);
 
-		if (pen.penDown && (pen.previousPoint.x != sprite.look.getX() || pen.previousPoint.y != sprite.look.getY())) {
+		PointF nextPoint = pen.pointsToDraw.poll();
+		while (nextPoint != null) {
 			renderer.circle(pen.previousPoint.x, pen.previousPoint.y, pen.penSize / 2);
-			renderer.rectLine(pen.previousPoint.x, pen.previousPoint.y, x, y, pen.penSize);
-			renderer.circle(x, y, pen.penSize / 2);
+			renderer.rectLine(pen.previousPoint.x, pen.previousPoint.y, nextPoint.x, nextPoint.y, pen.penSize);
+			renderer.circle(nextPoint.x, nextPoint.y, pen.penSize / 2);
+
+			pen.previousPoint = nextPoint;
+			nextPoint = pen.pointsToDraw.poll();
 		}
 
 		renderer.end();
-		pen.previousPoint.x = x;
-		pen.previousPoint.y = y;
 	}
 
-	public void dispose() {
+	public synchronized void dispose() {
```

Comment:
I can't say for sure, but I think this is the only method causing problems with concurrency.
The `StageListener.finish` method (which calls `PenActor.dispose`) is also called from other Threads. Could the `StageListener.dispose` handle it? Would that solve all concurrency problems?

Reply:
I looked into it, it should also be done when reloading the project. Might also cause issues when "reset" is called.
This is what I came up with to fix it for the release: https://github.com/Catrobat/Catroid/pull/2872

Reply:
What I meant, is, instead of disposing the `PenActor` in `StageListener.finish`, dispose it in `StageListener.dispose`. No need for synrchonization because all ApplicationListener methods are called from the OpenGL Thread.

> All methods are called in a thread that has the OpenGL context current. You can thus safely create and manipulate graphics resources.